Subject: [PATCH] gitweb: include navigation for non-commit-ishes

It's possible to view a tree or blob directly by using the tree
or blob's hash or a tag that points directly to one (without
an intervening commit).  In that case it does not make sense
to show 'commit', 'commitdiff' or 'tree' (in the case of a blob)
options in the navigation line, but things like 'summary' and
'log' still apply.

Modify the code to provide minimal navigation links in these
cases rather than nothing at all which was previously the case.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>
---
 gitweb/gitweb.perl | 31 ++++++++++++++++++++++---------
 1 file changed, 22 insertions(+), 9 deletions(-)

diff --git a/gitweb/gitweb.perl b/gitweb/gitweb.perl
index d678a11d..e8e020ae 100755
--- a/gitweb/gitweb.perl
+++ b/gitweb/gitweb.perl
@@ -4853,7 +4853,13 @@ sub git_print_page_nav {
 
 	my @navs = qw(summary log commit commitdiff tree refs);
 	if ($suppress) {
-		@navs = grep { $_ ne $suppress } @navs;
+		my %omit;
+		if (ref($suppress) eq 'ARRAY') {
+			%omit = map { ($_ => 1) } @$suppress;
+		} else {
+			%omit = ($suppress => 1);
+		}
+		@navs = grep { !$omit{$_} } @navs;
 	}
 
 	my %arg = map { $_ => {action=>$_} } @navs;
@@ -7434,10 +7440,19 @@ sub git_tag {
 	$fullhash = $hash if $hash =~ m/^[0-9a-fA-F]{40}$/;
 	$fullhash = git_get_full_hash($project, $hash) unless $fullhash;
 
-	my $head = git_get_head_hash($project);
+	my $obj = $tag{'object'};
 	git_header_html();
-	git_print_page_nav('','', $head,undef,$head);
-	git_print_header_div('commit', esc_html($tag{'name'}), $hash);
+	if ($tag{'type'} eq 'commit') {
+		git_print_page_nav('','', $obj,undef,$obj);
+		git_print_header_div('commit', esc_html($tag{'name'}), $hash);
+	} else {
+		if ($tag{'type'} eq 'tree') {
+			git_print_page_nav('',['commit','commitdiff'], undef,undef,$obj);
+		} else {
+			git_print_page_nav('',['commit','commitdiff','tree'], undef,undef,undef);
+		}
+		print "<div class=\"title\">".esc_html($hash)."</div>\n";
+	}
 	print "<div class=\"title_text\">\n" .
 	      "<table class=\"object_header\">\n" .
 	      "<tr><td>tag</td><td class=\"sha1\">$fullhash</td></tr>\n" .
@@ -7926,9 +7941,8 @@ sub git_blob {
 		git_print_page_nav('','', $hash_base,$co{'tree'},$hash_base, $formats_nav);
 		git_print_header_div('commit', esc_html($co{'title'}), $hash_base);
 	} else {
-		print "<div class=\"page_nav\">\n" .
-		      "<br/><br/></div>\n" .
-		      "<div class=\"title\">".esc_html($hash)."</div>\n";
+		git_print_page_nav('',['commit','commitdiff','tree'], undef,undef,undef);
+		print "<div class=\"title\">".esc_html($hash)."</div>\n";
 	}
 	git_print_page_path($file_name, "blob", $hash_base);
 	print "<div class=\"title_text\">\n" .
@@ -8015,9 +8029,8 @@ sub git_tree {
 		                   join(' | ', @views_nav));
 		git_print_header_div('commit', esc_html($co{'title'}), $hash_base, undef, $ref);
 	} else {
+		git_print_page_nav('tree',['commit','commitdiff'], undef,undef,$hash_base);
 		undef $hash_base;
-		print "<div class=\"page_nav\">\n";
-		print "<br/><br/></div>\n";
 		print "<div class=\"title\">".esc_html($hash)."</div>\n";
 	}
 	if (defined $file_name) {
---
